Detectors/EMCAL/workflow/src/PublisherSpec.cxx

```diff
@@ -0,0 +1,149 @@
+
+    auto processAttributes = std::make_shared<ProcessAttributes>();
+    {
+      processAttributes->terminateOnEod = ic.options().get<bool>("terminate-on-eod");
+      processAttributes->finished = false;
+      processAttributes->datatype = config.databranch.defval;
+      auto& dto = config.dataoutput;
+      auto& mco = config.mcoutput;
+      constexpr auto persistency = o2::framework::Lifetime::Timeframe;
+      o2::header::DataHeader::SubSpecificationType subSpec = 0;
+      if (propagateMC) {
+        processAttributes->reader = std::make_shared<o2::framework::RootTreeReader>(treename.c_str(), // tree name
+                                                                                    filename.c_str(), // input file name
+                                                                                    nofEvents,        // number of entries to publish
+                                                                                    publishingMode,
+                                                                                    o2::framework::Output{ dto.origin, dto.description, subSpec, persistency },
+                                                                                    dtbrName.c_str(), // name of cluster branch
+                                                                                    o2::framework::Output{ mco.origin, mco.description, subSpec, persistency },
+                                                                                    mcbrName.c_str() // name of mc label branch
+        );
+      } else {
+        processAttributes->reader = std::make_shared<o2::framework::RootTreeReader>(treename.c_str(), // tree name
+                                                                                    filename.c_str(), // input file name
+                                                                                    nofEvents,        // number of entries to publish
+                                                                                    publishingMode,
+                                                                                    o2::framework::Output{ dto.origin, dto.description, subSpec, persistency },
+                                                                                    dtbrName.c_str() // name of cluster branch
+        );
+      }
+    }
+
+    auto processFunction = [processAttributes, propagateMC](o2::framework::ProcessingContext& pc) {
+      if (processAttributes->finished) {
+        return;
+      }
+
+      auto publish = [&processAttributes, &pc, propagateMC]() {
+        o2::emcal::EMCALBlockHeader::PayloadType_t payloadType(o2::emcal::EMCALBlockHeader::PayloadType_t::kNoPayload);
+        LOG(DEBUG) << "[publish] Data type:" << processAttributes->datatype;
+        if (processAttributes->datatype.find("EMCALDigit") != std::string::npos) {
```

Comment:
I would consider this if statement a pure debug feature which should be removed. Also, there is no need to propagate the payload type in a custom header because the data header is intended for this information. What you have to make sure in the invocation of the `getPublisherConf` method is to use the correct DataDescription.

Edit: I have now realized that you are using the payload type of the EMCALBlockHeader to propagate the end-of-data condition. I suggest to rename it, and we will find a generic solution in DPL.

Reply:
Debug printf will be removed before merging. For the field in the header I move it to simple bool.  For the moment we need the header, otherwise components in particular in the QC will break.